### PR TITLE
Add option to use underscore symbol `_` instead of `*` to define anonymous type lambdas

### PR DESCRIPTION
**Pull request by neko-kai**
== Dev Notes

This PR depends on the following scalac changes:

- https://github.com/scala/scala/pull/9605 
- https://github.com/scala/scala/pull/9614 

To be widely applicable (without these changes the user would not have a way to write wildcards when using this mode and would have to write +_ and -_ with backticks – both would hinder cross-compilation with Scala 3)

The tests pass on all scalac versions (I've quoted `+_`, `-_` for 2.10/2.11) compat. I've found the built-in test suite to be comprehensive, but small, so I've made sure that this change works by porting https://github.com/7mind/izumi/pull/1479 to the new syntax, with unquoted `+_`/`-_` and, well, it works on my machine.

Fixes https://github.com/typelevel/kind-projector/issues/120

== Description

The syntax roughly follows the [proposed new syntax for wildcards and placeholders](https://dotty.epfl.ch/docs/reference/changed-features/wildcards.html#migration-strategy) for Scala 3.2+ and is designed to allow cross-compilation of libraries between Scala 2 and Scala 3 while using the new Scala 3 syntax for both versions.

To enable this mode, add `-P:kind-projector:underscore-placeholders` to your scalac command-line. In sbt you may do this as follows:

```scala
ThisBuild / scalacOptions += "-P:kind-projector:underscore-placeholders"
```

This mode is designed to be used with scalac versions `2.12.14`+ and `2.13.6`+, these versions add an the ability to use `?` as the existential type wildcard ([scala/scala#9560](https://github.com/scala/scala/pull/9560)), allowing to repurpose the underscore without losing the ability to write existential types. It is not advised that you use this mode with older versions of scalac or without `-Xsource:3` flag, since you will lose the underscore syntax entirely.

Here are a few examples:

```scala
Tuple2[_, Double]        // equivalent to: type R[A] = Tuple2[A, Double]
Either[Int, +_]          // equivalent to: type R[+A] = Either[Int, A]
Function2[-_, Long, +_]  // equivalent to: type R[-A, +B] = Function2[A, Long, B]
EitherT[_[_], Int, _]    // equivalent to: type R[F[_], B] = EitherT[F, Int, B]
```

Examples with `-Xsource:3`'s `?`-wildcard:

```scala
Tuple2[_, ?]        // equivalent to: type R[A] = Tuple2[A, x] forSome { type x }
Either[?, +_]          // equivalent to: type R[+A] = Either[x, A] forSome { type x }
Function2[-_, ?, +_]  // equivalent to: type R[-A, +B] = Function2[A, x, B] forSome { type x }
EitherT[_[_], ?, _]    // equivalent to: type R[F[_], B] = EitherT[F, x, B] forSome { type x }
```

### CODE REVIEW

File: build.sbt
Comment (larsrh) [inline build.sbt:88]: Do we need the Scala 2.10 support still? Otherwise I'm inclined to drop this.

Comment (neko-kai) [reply build.sbt:88]: @larsrh Do you want me to drop 2.10 build as part of this PR? The `case` is required to compile on 2.10 but is not required otherwise.

Comment (larsrh) [reply build.sbt:88]: That would be great!

Comment (neko-kai) [reply build.sbt]: @larsrh Done in last 2 commits

**Comment by larsrh** [comment]
I can't push to your branch, but you'd have to run `sbt githubWorkflowGenerate` for CI to pass.

**Review by larsrh — APPROVED** [review]
LGTM, pending green build

**Comment by neko-kai** [comment]
@larsrh Build's green now 👍 

> I can't push to your branch

Odd, I don't think I changed anything to prohibit this, though I do not see an 'allow edits from maintainers' checkbox on this page either.

**Comment by larsrh** [comment]
Thanks!

**Comment by larsrh** [comment]
Since this is a new feature, not a breaking change, I think we can roll this as 0.12.1. I will update the README accordingly.

**Comment by neko-kai** [comment]
@larsrh Thanks! :pray: For the release number, everywhere in the docs for the new feature I wrote version `0.13.0 and up` (migration guide, dotty wildcards page, readme). It could be easier to release as 0.13 just to not update all of these...

**Comment by larsrh** [comment]
Fair enough running that now.

**Pull request merged**
